File: backend/models/agency_assessment.py
from models import db


class AgencyAssessment(db.Model):
    """代理机构服务质量考核（一个项目一份）。

    对应客户《招标代理机构服务质量考核评价表》：
      二、考核内容及评分标准 —— 15 个加/扣分项，满分 100，得分 = 100 + 各项加扣分之和
      三、一票否决项目 —— 9 条，命中即一票否决
      四、综合评价 —— 3 项主观勾选 + 文字意见（非评分项）

    其中 6 项能从 PMS 已有数据自动算出建议分（时效类靠时间戳、规范类靠驳回留痕），
    系统给建议值，采购部对接人可改——机器算得出的不让人再算一遍，算不出的才靠人。
    """
    __tablename__ = "agency_assessments"

    id           = db.Column(db.Integer, primary_key=True)
    project_id   = db.Column(db.Integer, index=True, nullable=False)
    project_number = db.Column(db.String(50), default="")
    project_name = db.Column(db.String(200), default="")
    agency_code  = db.Column(db.String(10), default="", index=True)
    agency_name  = db.Column(db.String(100), default="")

    # 15 个评分项：[{key, score, note, auto_score, auto_basis}]
    items_json   = db.Column(db.Text, default="[]")
    # 一票否决：命中的条目 key 列表
    veto_json    = db.Column(db.Text, default="[]")
    veto_note    = db.Column(db.Text, default="")

    # 综合评价（非评分项）：满意 | 一般 | 不满意
    subj_timeliness = db.Column(db.String(10), default="")   # 经办人响应的及时性
    subj_ability    = db.Column(db.String(10), default="")   # 经办人水平和能力
    subj_attitude   = db.Column(db.String(10), default="")   # 合作态度及协调能力
    comment      = db.Column(db.Text, default="")            # 建议或意见

    total_score  = db.Column(db.Float, default=100.0)        # 100 + 各项加扣分之和
    veto_hit     = db.Column(db.Integer, default=0)          # 1=触发一票否决

    status       = db.Column(db.String(10), default="草稿")   # 草稿 | 已提交
    assessor     = db.Column(db.String(50), default="")      # 采购部对接人
    assessed_at  = db.Column(db.String(30), default="")
    created_by   = db.Column(db.String(50), default="")
    created_at   = db.Column(db.String(30), default="")
    updated_at   = db.Column(db.String(30), default="")

    def to_dict(self):
        import json
        d = {c.name: getattr(self, c.name) for c in self.__table__.columns}
        try:
            d["items"] = json.loads(self.items_json or "[]")
        except Exception:
            d["items"] = []
        try:
            d["veto"] = json.loads(self.veto_json or "[]")
        except Exception:
            d["veto"] = []
        return d

File: frontend/src/services/agencyAssessment.ts
import api from './api'

export interface AssessItem {
  key: string
  name: string
  standard: string
  auto: boolean
  auto_score: number | null   // 系统算出的建议分（null=算不出）
  auto_basis: string          // 建议分的依据，写清楚为什么是这个分
  score: number
  note: string
  // 三项时效（编制/拟合同/归档）才有：起止日期可手填，填了就按手填的算分
  date_start?: string
  date_end?: string
  date_source?: 'manual' | 'auto' | 'none'
  start_label?: string
  end_label?: string
  auto_hint?: string
}

// {"archive_speed": {"start": "2026-06-01", "end": "2026-06-03"}, ...}
export type AssessDates = Record<string, { start?: string; end?: string }>

export interface VetoItem { key: string; name: string }

export interface Assessment {
  id: number | null
  project_id: number
  project_number: string
  project_name: string
  agency_code: string
  agency_name: string
  items: AssessItem[]
  veto: string[]
  veto_note: string
  dates?: AssessDates
  veto_hit?: number
  subj_timeliness: string
  subj_ability: string
  subj_attitude: string
  comment: string
  total_score: number
  status: '草稿' | '已提交'
  assessor?: string
  assessed_at?: string
}

export interface AssessMeta {
  items: { key: string; name: string; standard: string; auto: boolean }[]
  veto_items: VetoItem[]
  subj_options: string[]
  ladder_items?: Record<string, { start_label: string; end_label: string; auto_hint: string }>
  thresholds: { pass_line: number; bonus_line: number; suspend_line: number; valid_months: number }
  can_assess: boolean
}

export interface AgencySummary {
  agency_code: string
  agency_name: string
  count: number
  avg: number | null
  net: number          // 近 N 月加扣分净额
  veto: number
  below_count?: number
  flags: string[]
  advice: string
  months?: number
  period?: string      // 这份汇总统计的是哪一段时间，直接显示给人看
}

export interface PendingProject {
  id: number; number: string; name: string; method: string
  officer: string; agency_code: string; agency_name: string; status: string
}

export const getAssessMeta = () =>
  api.get<{ ok: boolean; data: AssessMeta }>('/agency-assessments/meta')

export const listAssessments = (params?: Record<string, string>) =>
  api.get<{ ok: boolean; data: Assessment[] }>('/agency-assessments', { params })

export const listPendingProjects = () =>
  api.get<{ ok: boolean; data: PendingProject[] }>('/agency-assessments/pending-projects')

export const getAssessment = (pid: number) =>
  api.get<{ ok: boolean; data: Assessment }>(`/agency-assessments/project/${pid}`)

export const saveAssessment = (pid: number, data: Partial<Assessment>, submit = false) =>
  api.post<{ ok: boolean; data: Assessment; message: string }>(
    `/agency-assessments/project/${pid}${submit ? '?submit=1' : ''}`, data)

export const revokeAssessment = (aid: number) =>
  api.post<{ ok: boolean; message: string }>(`/agency-assessments/${aid}/revoke`)

// start/end 是「2026-01」这样的月份，含起含止；都不传就是考核办法默认的近 3 个月
export const getAgencySummary = (start?: string, end?: string) =>
  api.get<{ ok: boolean; data: AgencySummary[]; period?: string }>(
    '/agency-assessments/summary', { params: { start, end } })

// 只重算不落库：填日历时用它拿到新的建议分，算分规则不在前端复制一份
export const previewAssessment = (pid: number, items: AssessItem[], dates: AssessDates) =>
  api.post<{ ok: boolean; data: { items: AssessItem[]; total_score: number } }>(
    `/agency-assessments/project/${pid}/preview`, { items, dates })

// 导出 Excel / 打印：后端出成稿，走浏览器直连（要带 cookie，所以不用 axios）
export const assessExportUrl = (pid: number) =>
  `${api.defaults.baseURL}/agency-assessments/project/${pid}/export.xlsx`
export const assessPrintUrl = (pid: number) =>
  `${api.defaults.baseURL}/agency-assessments/project/${pid}/print`
